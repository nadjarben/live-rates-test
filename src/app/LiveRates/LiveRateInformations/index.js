import React from "react";
import { convertFormatDate } from "../../../helpers";

const LiveRateInformations = ({
  liveRate: { 
    currency, 
    rate, 
    bid, 
    ask, 
    low, 
    high, 
    open, 
    close, 
    timestamp 
  },
}) => {
  const relativeDate = React.useCallback(() => {
    return convertFormatDate(timestamp)
  }, [timestamp])

  return (
    <tr>
      <td>{currency}</td>
      <td>{rate}</td>
      <td>{bid}</td>
      <td>{ask}</td>
      <td>{high}</td>
      <td>{low}</td>
      <td>{open}</td>
      <td>{close}</td>
      <td>{relativeDate()}</td>
    </tr>
  );
};

export default LiveRateInformations;
