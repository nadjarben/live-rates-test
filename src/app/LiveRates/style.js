import styled from "styled-components";

const Wrapper = styled.div`
  position: relative;

  & > h1 {
    text-align: center;
  }

  & > h3 {
    text-align: center;
    color: red;
  }

  & > table {
    border: 1px solid #ddd;
    padding: 8px;
    width: 100%;
  }

  th {
    padding-top: 15px;
    padding-bottom: 15px;
    text-align: center;
    background-color: #5652BF;
    color: white;
  }

  tr {
    text-align: center;
  }

  tr:nth-child(even) {
    background-color: #f2f2f2;
  }

  tr:hover {
    background-color: #ddd;
  }

  td {
    padding-top: 10px;
    padding-bottom: 10px;
  }
`;

export { Wrapper };
