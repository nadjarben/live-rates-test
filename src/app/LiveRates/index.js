import React from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  fetchLiveRates,
  updateLiveRates,
} from "../../store/actions/liveRates.action";
import LiveRateInformations from "./LiveRateInformations";
import Loader from "../../components/Loader";
import Nofication from "../../components/Notification";
import { Wrapper } from "./style";

const LiveRates = () => {
  const dispatch = useDispatch();
  const liveRates = useSelector((state) => state.liveRates);

  React.useEffect(() => {
    dispatch(fetchLiveRates());
  }, [dispatch]);

  React.useEffect(() => {
    const interval = setInterval(() => {
      dispatch(updateLiveRates());
      return () => clearInterval(interval);
    }, 10000);
  }, [dispatch]);

  const { liveRatesList, isLoading, errorMessage } = liveRates;

  const renderTableHeader = React.useCallback(() => {
    let header = Object.keys(liveRatesList[0]);
    return header.map((key, index) => (
      <th key={index}>
        {key !== "timestamp" ? key.toUpperCase() : "LAST UPDATED"}
      </th>
    ));
  }, [liveRatesList]);

  return (
    <Wrapper>
      <h1>Live Rates</h1>
      {errorMessage && <Nofication isError message={errorMessage} />}
      {isLoading ? (
        <Loader />
      ) : (
        <table>
          <thead>
            <tr>{liveRatesList[0] && renderTableHeader()}</tr>
          </thead>
          <tbody>
            {liveRatesList.map((liveRate, index) => (
              <LiveRateInformations key={index} liveRate={liveRate} />
            ))}
          </tbody>
        </table>
      )}
    </Wrapper>
  );
};

export default LiveRates;
