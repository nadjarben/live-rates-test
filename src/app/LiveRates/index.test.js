import React from "react";
import { Provider } from "react-redux";
import { store } from "../../store";
import LiveRates from "./";


const ShallowLiveRates = (
  <Provider store={store}>
    <LiveRates />
  </Provider>
);

describe("LiveRates", () => {
  test("Does match snapshots", () => {
    expect(ShallowLiveRates).toMatchSnapshot();
  });

  test("Does LiveRates be defined", () => {
    expect(ShallowLiveRates).toBeDefined();
  });
});
