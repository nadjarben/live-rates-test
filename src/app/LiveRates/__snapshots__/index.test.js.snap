// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveRates Does match snapshots 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <LiveRates />
</Provider>
`;
